crates/searcher/src/searcher/mod.rs

```diff
@@ -790,6 +797,16 @@ impl Searcher {
         self.config.encoding.is_some()
             || (self.config.bom_sniffing && slice_has_utf16_bom(slice))
```

Comment:
I think it would be better to implement this the same way that UTF-16 is. Namely, tweak this to:

```
(self.config.bom_sniffing && slice_has_bom(slice))
```

and then create `slice_has_bom` by merging `slice_has_utf16_bom` and `slice_has_utf8_bom`. And then remove the `slice_has_utf8_bom_to_sniff` helper that you created and the corresponding check in `search_slice`. The `search_reader` implementation should then handle it.

The main problem here is that UTF-8 BOM stripping only _sometimes_ doesn't work. Namely, it doesn't work in precisely the case where ripgrep searches via memory maps. That's why, if a UTF-16 BOM is detected, then it falls back to `search_reader` because `search_reader` will do all the necessary transcoding and BOM stripping.

e.g., looking at #1638, while `rg foo utf8bom --json` does indeed show the bug, `rg foo utf8bom --json --no-mmap` does not, because `search_reader` gets this right. So I think we should just divert to that like we do for UTF-16.

Reply:
Thanks for reviewing this! I'll push those changes you requested shortly.

https://github.com/BurntSushi/ripgrep/blob/def993bad1a9275cdc249f04048e5b2065b79f05/crates/searcher/src/searcher/mod.rs#L748-L752
When implementing the fix, this comment on line 748 implies that `search_reader` doesn't have the same perfomance that the other methods have. That's why I implemented UTF-8 BOM handling after `search_reader` option is discarded. But indeed the code will look much cleaner with all BOM sniffing done by `search_reader`.

Reply:
https://github.com/BurntSushi/ripgrep/blob/def993bad1a9275cdc249f04048e5b2065b79f05/crates/searcher/src/searcher/mod.rs#L976-L980
The same performance warning is mentioned on `slice_has_utf16_bom`

Reply:
Yes. Technically, mmaps can be a bit faster in some cases. But UTF-8 BOM's are quite rare, and it's not worth the code complexity IMO.